### PR TITLE
Date accessors

### PR DESCRIPTION
**Pull request by cbeer**

### CODE REVIEW

File: lib/mods/date.rb [hunk 1]
```diff
@@ -61,11 +68,11 @@ def self.cleanup(text)
...
     # MARC-formatted date parser, similar to EDTF, but with special support for
     # MARC-specific encodings
-    class MarcFormat < EdtfFormat
+    class MarcFormat < Date
       def self.cleanup(text)
-        return nil if text == "9999" || text == "uuuu"
+        return nil if text == "9999" || text == "uuuu" || text == '||||'
 
-        text.gsub(/^[\[]+/, '').gsub(/[\.\]]+$/, '')
+        Date.cleanup(text)
```
Comment (jcoyne) [inline lib/mods/date.rb]: Can you call `super` here?

File: lib/mods/date.rb [hunk 2]
```diff
@@ -95,11 +102,11 @@ def self.supports?(text)
 
     # Full text extractor for MM/DD/YYYY-formatted dates
```
Comment (jcoyne) [inline lib/mods/date.rb]: Does this comment need to change? Are we accepting 3 digit years now?

File: lib/mods/date.rb [hunk 3]
```diff
@@ -124,15 +131,15 @@ def self.munge_to_yyyy(text)
...
+      REGEX = /(?<![A-Za-z\.])(?<year>[MCDLXVI\.]+)(?![A-Za-z])/
 
       def self.cleanup(text)
         matches = text.match(REGEX)
         roman_to_int(matches[:year].upcase).to_s
       end
 
       def self.roman_to_int(value)
-        value = value.dup
+        value = value.dup.tr('.', '')
```
Comment (jcoyne) [inline lib/mods/date.rb]: Do we need to dup if we're running a `tr` won't that create a copy for you?

File: lib/mods/date.rb [hunk 4]
```diff
@@ -145,9 +152,10 @@ def self.roman_to_int(value)
       end
     end
 
-    # Full-text extractor for centuries encoded as Roman numerals
+    # Full-text extractor for centuries encoded as Roman numerals; sometimes centuries
+    # are given as e.g. xvith, hence the funny negative look-ahead assertion
```
Comment (jcoyne) [inline lib/mods/date.rb]: 👍

File: lib/mods/date.rb [hunk 5]
```diff
@@ -181,16 +189,77 @@ def self.cleanup(text)
       end
     end
 
+    # Full-text extractor for data formatted as YYYY-YYYY
```
Comment (jcoyne) [inline lib/mods/date.rb]: or `YYY-YYY`?

File: lib/mods/date.rb [hunk 6]
```diff
@@ -181,16 +189,77 @@ def self.cleanup(text)
...
+    class YearRangeFormat < ExtractorDateFormat
+      REGEX = /(?<start>\d{3,4})-(?<end>\d{3,4})/
+
+      def self.cleanup(text)
```
Comment (jcoyne) [inline lib/mods/date.rb]: Add a comment that this normalizes to `YYYY/YYYY`

File: lib/mods/date.rb [hunk 7]
```diff
@@ -181,16 +189,77 @@ def self.cleanup(text)
...
+      end
+    end
+
+    # Full-text extractor for data formatted as YYY-
+    class DecadeAsYearDashFormat < ExtractorDateFormat
+      REGEX = /(?<year>\d{3})[-_x?](?!\d)/
+
+      def self.cleanup(text)
+        matches = text.match(REGEX)
+        "#{matches[:year]}X"
```
Comment (jcoyne) [inline lib/mods/date.rb]: What does this mean? Can you leave a comment?

File: lib/mods/date.rb [hunk 8]
```diff
@@ -181,16 +189,77 @@ def self.cleanup(text)
...
+      end
+    end
+
+    # Full-text extractor that tries hard to pick any year present in the data
+    class EmbeddedBCYearFormat < ExtractorDateFormat
+      REGEX = /(?<year>\d{3,4})\s?B\.?C\.?/i
+
+      def self.cleanup(text)
```
Comment (jcoyne) [inline lib/mods/date.rb]: add comment that this normalizes to `-YYYY`

File: lib/mods/date.rb [hunk 9]
```diff
@@ -181,16 +189,77 @@ def self.cleanup(text)
...
+        matches = text.match(REGEX)
+        "-#{(matches[:year].to_i - 1).to_s.rjust(4, "0")}"
+      end
+    end
+
     # Full-text extractor that tries hard to pick any year present in the data
     class EmbeddedYearFormat < ExtractorDateFormat
-      REGEX = /(?<prefix>-)?(?<year>\d{3,4})/
+      REGEX = /(?<prefix>-)?(?<!\d)(?<year>\d{4})(?!\d)/
```
Comment (jcoyne) [inline lib/mods/date.rb]: Can you give an example of what a prefix might be?

File: lib/mods/date.rb [hunk 10]
```diff
@@ -181,16 +189,77 @@ def self.cleanup(text)
...
+
+      def self.cleanup(text)
+        matches = text.match(REGEX)
+        "#{matches[:prefix]}#{matches[:year].rjust(4, "0")}"
+      end
+    end
+
+    # Full-text extractor that tries hard to pick any year present in the data
+    class EmbeddedYearWithBracketsFormat < ExtractorDateFormat
+      # [YYY]Y Y[YYY] [YY]YY Y[YY]Y YY[YY] YYY[Y] YY[Y]Y Y[Y]YY [Y]YYY
```
Comment (jcoyne) [inline lib/mods/date.rb]: Woah. 😱  Can you add a comment about what this stuff means?  Is there a part of a spec I can look at?

Comment (cbeer) [reply lib/mods/date.rb]: No, I have no idea where it comes from, I just know we have lots of it.

File: lib/mods/date.rb (hunk 9 above)
Comment (cbeer) [reply lib/mods/date.rb]: A literal `-`?

File: lib/mods/date.rb (hunk 8 above)
Comment (cbeer) [reply lib/mods/date.rb]: I will rename `cleanup` to `normalize_to_edtf` for clarity.

File: lib/mods/date.rb (hunk 9 above)
Comment (jcoyne) [reply lib/mods/date.rb]: 👍

**Pull request merged**